### ISSUE TITLE
error: The option `virtualisation.docker.storageDriver' is used but not defined.

### ISSUE THREAD

**Issue by bjornfor** [NixOS/nixpkgs#9801]
I'm getting this error on latest release-15.09 channel (15.09.308.2734724):
```
$ nixos-rebuild build -Q
building Nix...
building the system configuration...
error: The option `virtualisation.docker.storageDriver' is used but not defined.
(use ‘--show-trace’ to show detailed location information)
```

I'm guessing this commit caused it: https://github.com/NixOS/nixpkgs/commit/c2d1617b91ed4bb0b589d133da7f5428dd625b12

**Comment by bjornfor** [comment]
I just read the option documentation for virtualsation.docker.storageDriver, which explains that there is (deliberately) no default value for it.

IMHO, it would be more user friendly to use and assert. (Is it really no reasonable default value for this option? Not even a value that comes from another part of the nixos config?)

**Comment by domenkozar** [comment]
+1 for assert (and we should document how values without defaults are handled in nixos manual)

**Comment by bjornfor** [comment]
I tried this:

```
diff --git a/nixos/modules/virtualisation/docker.nix b/nixos/modules/virtualisation/docker.nix
index ba078cc..20f06b5 100644
--- a/nixos/modules/virtualisation/docker.nix
+++ b/nixos/modules/virtualisation/docker.nix
@@ -45,7 +45,8 @@ in
       };
     storageDriver =
       mkOption {
-        type = types.enum ["aufs" "btrfs" "devicemapper" "overlay" "zfs"];
+        type = types.nullOr (types.enum ["aufs" "btrfs" "devicemapper" "overlay" "zfs"]);
+        default = null;
         description =
           ''
             This option determines which Docker storage driver to use.
@@ -86,6 +87,14 @@ in
   ###### implementation
 
   config = mkIf cfg.enable (mkMerge [
+    { assertions = [ {
+        assertion = cfg.storageDriver != null;
+        message = ''
+          Option services.docker.storageDriver is not defined (required because
+          services.docker.enable is true).
+        '';
+        } ];
+    }
     { environment.systemPackages = [ pkgs.docker ];
       users.extraGroups.docker.gid = config.ids.gids.docker;
     }
```

But got "error: cannot coerce null to a string". Is it because the mkMerge causes the option to be evaluated before the assert?

**Comment by bjornfor** [comment]
I just read the upstream documentation: https://docs.docker.com/reference/commandline/daemon/ (search for "Daemon storage-driver option"). At first the "overlay" option seemed like a good default, but they don't recommend it in production (see above link).

**Comment by bjornfor** [comment]
Hm, what is the default storage-driver option that docker uses if nothing is specified on the command-line? (I couldn't find it in the docker documentation.)

I find it odd that NixOS forces users to decide on something that docker itself does not.

CC @ragnard.

**Comment by Mayeu** [comment]
I believe the default is `devicemapper`, was that on Archlinux, and from my previously working NixOS docker I have the devicemapper folder by default:
```
$ sudo ls /var/lib/docker 
containers
devicemapper
graph
init
linkgraph.db
repositories-devicemapper
tmp
trust
vfs
volumes
```

**Comment by listx** [comment]
FWIW I also have devicemapper as the default storage driver on NixOS.

**Comment by bjornfor** [comment]
Fixed in master and release-15.09.
